File: kickstart/2020-B/robot_path.py
T = int(input())
for t in range(T):
    program = input()

    mults = []
    diffs = [(0, 0)]

    for c in program:
        if c == "(":
            diffs.append((0, 0))
        elif c == ")":
            m = mults.pop()
            dx, dy = diffs.pop()
            x, y = diffs[-1]
            x += dx * m
            y += dy * m
            diffs[-1] = (x, y)
        elif c.isdigit():
            mults.append(int(c))
        else:
            x, y = diffs[-1]
            if c == "N":
                y -= 1
            elif c == "S":
                y += 1
            if c == "E":
                x += 1
            if c == "W":
                x -= 1
            diffs[-1] = (x, y)

    x, y = diffs[-1]
    x = 1 + (x % 1000000000)
    y = 1 + (y % 1000000000)
    print("Case #{}: {} {}".format(t + 1, x, y))

File: kickstart/2020-B/robot_path.cpp
#include <bits/stdc++.h>
using namespace std;

long long bound(long long x) {
  if (x > 1000000000) {
    return 1000000000 - (x % 1000000000);
  } else if (x <= 0) {
    return 1000000000 + (x % 1000000000);
  } else {
    return x;
  }
}

int main() {
  long long T;
  cin >> T;
  for (long long t = 0; t < T; t++) {
    string program;
    cin >> program;

    stack<pair<long long, long long>> diffs;
    stack<long long> mults;
    diffs.push(pair<long long, long long>(1, 1));

    for (char &c : program) {
      if (c == '(')
        diffs.push(pair<long long, long long>(0, 0));
      else if (c == ')') {
        long long m = mults.top();
        mults.pop();

        pair<long long, long long> diff = diffs.top();
        long long dx = diff.first;
        long long dy = diff.second;
        diffs.pop();

        pair<long long, long long> top = diffs.top();
        diffs.pop();
        top.first = (top.first + dx * m) % 1000000000;
        top.second = (top.second + dy * m) % 1000000000;
        diffs.push(top);
      } else if (isdigit(c)) {
        mults.push((int)c - 48);
      } else {
        auto top = diffs.top();
        diffs.pop();

        switch (c) {
        case 'N':
          top.second = (top.second - 1);
          break;
        case 'S':
          top.second = (top.second + 1);
          break;
        case 'E':
          top.first = (top.first + 1);
          break;
        case 'W':
          top.first = (top.first - 1);
          break;
        }
        diffs.push(top);
      }
    }

    long long x = bound(diffs.top().first);
    long long y = bound(diffs.top().second);
    cout << "Case #" << t + 1 << ": " << x << " " << y << endl;
  }
}
